### PR TITLE
[SPARK-20055] [Docs] Added documentation for loading csv files into DataFrames

### PR DESCRIPTION
**Pull request by jomach**
## What changes were proposed in this pull request?

 Added documentation for loading csv files into Dataframes

## How was this patch tested?

/dev/run-tests

### CODE REVIEW

**Review by felixcheung — COMMENTED** [review]
there is no `people.csv` though. these examples need to be runnable

**Comment by jomach** [comment]
@felixcheung  Sorry for that. Should be there now. Can you test ? thanks

File: examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java [hunk 1]
```diff
@@ -115,7 +115,20 @@ private static void runBasicDataSourceExample(SparkSession spark) {
     Dataset<Row> peopleDF =
       spark.read().format("json").load("examples/src/main/resources/people.json");
     peopleDF.select("name", "age").write().format("parquet").save("namesAndAges.parquet");
-    // $example off:manual_load_options$
+    // $example on:manual_load_options_csv$
```
Comment (gatorsmile) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java]: You still need to keep 
> // $example off:manual_load_options$

File: examples/src/main/python/sql/datasource.py [hunk 2]
```diff
@@ -53,6 +53,11 @@ def basic_datasource_example(spark):
     df.select("name", "age").write.save("namesAndAges.parquet", format="parquet")
     # $example off:manual_load_options$
 
+    # $example on:manual_load_options_csv$
+    df = spark.read.load("examples/src/main/resources/people.csv",
+                         format="csv", sep=":", inferSchema="true", header="true")
+    # $example off:manual_load_options_csv
```
Comment (gatorsmile) [inline examples/src/main/python/sql/datasource.py]: This need to be corrected to 
> # $example off:manual_load_options_csv$

File: examples/src/main/resources/people.csv [hunk 3]
```diff
@@ -0,0 +1,3 @@
+name;age;job
+Jorge;30;Developer
+Bob;32;Developer
```
Comment (gatorsmile) [inline examples/src/main/resources/people.csv]: Add an empty line.

File: examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java [hunk 4]
```diff
@@ -115,7 +115,20 @@ private static void runBasicDataSourceExample(SparkSession spark) {
...
+              .option("header", "true")
+              .load("examples/src/main/resources/people.csv");
+    // $example off:manual_load_options_csv$
+    // $example on:manual_load_options_csv$
+    Dataset<Row> peopleDFCsv = spark.read().format("csv")
+              .option("sep", ";")
+              .option("inferSchema", "true")
+              .option("header", "true")
+              .load("examples/src/main/resources/people.csv");
+    // $example off:manual_load_options_csv$
```
Comment (gatorsmile) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java]: Line 125-131 is a duplicate.

File: docs/sql-programming-guide.md [hunk 5]
```diff
@@ -479,6 +481,47 @@ source type can be converted into other types using this syntax.
...
+<div data-lang="python"  markdown="1">
+{% include_example manual_load_options_csv python/sql/datasource.py %}
+</div>
+
+<div data-lang="r"  markdown="1">
+{% include_example manual_load_options_csv r/RSparkSQLExample.R %}
+</div>
+</div>
+
+To load a csv file you can use:
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: This is also a duplicate.

**Comment by jomach** [comment]
@gatorsmile  I dressed your comments. Still I cannot use the jekyll build...
`SKIP_API=1 jekyll build --incremental
Configuration file: /Users/jorge/Downloads/spark/docs/_config.yml
       Deprecation: The 'gems' configuration option has been renamed to 'plugins'. Please update your config file accordingly.
            Source: /Users/jorge/Downloads/spark/docs
       Destination: /Users/jorge/Downloads/spark/docs/_site
 Incremental build: enabled
      Generating... 
  Liquid Exception: invalid byte sequence in US-ASCII in _layouts/redirect.html
`

File: examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java [hunk 6]
```diff
@@ -116,6 +116,13 @@ private static void runBasicDataSourceExample(SparkSession spark) {
       spark.read().format("json").load("examples/src/main/resources/people.json");
     peopleDF.select("name", "age").write().format("parquet").save("namesAndAges.parquet");
     // $example off:manual_load_options$
+    // $example on:manual_load_options_csv$
+    Dataset<Row> peopleDFCsv = spark.read().format("csv")
+              .option("sep", ";")
+              .option("inferSchema", "true")
+              .option("header", "true")
```
Comment (gatorsmile) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java]: Could you change the indents of line 121-123 to 2 spaces?

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala [hunk 7]
```diff
@@ -49,6 +49,14 @@ object SQLDataSourceExample {
     val peopleDF = spark.read.format("json").load("examples/src/main/resources/people.json")
     peopleDF.select("name", "age").write.format("parquet").save("namesAndAges.parquet")
     // $example off:manual_load_options$
+    // $example on:manual_load_options_csv$
+    val peopleDFCsv = spark.read.format("csv")
+         .option("sep", ";")
+         .option("inferSchema", "true")
+         .option("header", "true")
+         .load("examples/src/main/resources/people.csv")
```
Comment (gatorsmile) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: Could you change the indents of line 54-57 to 2 spaces?

**Comment by gatorsmile** [comment]
ok to test

**Comment by jomach** [comment]
@gatorsmile pr comments fixed. Sorry but is my first time.

File: docs/sql-programming-guide.md [hunk 8]
```diff
@@ -479,6 +481,25 @@ source type can be converted into other types using this syntax.
...
+</div>
+
+<div data-lang="python"  markdown="1">
+{% include_example manual_load_options_csv python/sql/datasource.py %}
+</div>
+
+<div data-lang="r"  markdown="1">
+{% include_example manual_load_options_csv r/RSparkSQLExample.R %}
+</div>
+</div>
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Let's add another newline here. It breaks rendering.

File: docs/sql-programming-guide.md [hunk 9]
```diff
@@ -461,6 +461,8 @@ name (i.e., `org.apache.spark.sql.parquet`), but for built-in sources you can al
 names (`json`, `parquet`, `jdbc`, `orc`, `libsvm`, `csv`, `text`). DataFrames loaded from any data
 source type can be converted into other types using this syntax.
 
+To load a json file you can use:
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: I'd say `JSON` instead of `json`.

File: examples/src/main/r/RSparkSQLExample.R [hunk 10]
```diff
@@ -112,6 +112,11 @@ namesAndAges <- select(df, "name", "age")
 write.df(namesAndAges, "namesAndAges.parquet", "parquet")
 # $example off:manual_load_options$
 
+# $example on:manual_load_options_csv$
```
Comment (HyukjinKwon) [inline examples/src/main/r/RSparkSQLExample.R]: I'd add a newline here above to keep consistent in this file

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala [hunk 11]
```diff
@@ -49,6 +49,14 @@ object SQLDataSourceExample {
     val peopleDF = spark.read.format("json").load("examples/src/main/resources/people.json")
     peopleDF.select("name", "age").write.format("parquet").save("namesAndAges.parquet")
     // $example off:manual_load_options$
+    // $example on:manual_load_options_csv$
+    val peopleDFCsv = spark.read.format("csv")
+	  .option("sep", ";")
```
Comment (HyukjinKwon) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: double-spaced (no tab of course ..)

File: examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java [hunk 12]
```diff
@@ -116,6 +116,13 @@ private static void runBasicDataSourceExample(SparkSession spark) {
       spark.read().format("json").load("examples/src/main/resources/people.json");
     peopleDF.select("name", "age").write().format("parquet").save("namesAndAges.parquet");
     // $example off:manual_load_options$
+    // $example on:manual_load_options_csv$
+    Dataset<Row> peopleDFCsv = spark.read().format("csv")
+	  .option("sep", ";")
```
Comment (HyukjinKwon) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaSQLDataSourceExample.java]: ditto

**Comment by HyukjinKwon** [comment]
When I opened a JIRA, I thought a chapter such as https://spark.apache.org/docs/latest/sql-programming-guide.html#json-datasets. This chapter, `Manually Specifying Options`, looks describing how to specify options BTW.

File: docs/sql-programming-guide.md [hunk 13]
```diff
@@ -479,6 +481,25 @@ source type can be converted into other types using this syntax.
 </div>
 </div>
 
+To load a csv file you can use:
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: ditto

**Comment by jomach** [comment]
@gatorsmile pr comments fixed.  The problem with the actual docs is that people wen start with spark usually don't start with JSON files but with CSV files to "see" something....

**Comment by HyukjinKwon** [comment]
I intended to explain `multiLine`, `inferSchema` and `header` which are quite arguably commonly used rather than just show up the examples. JSON one explains `multiLine` and each line of the examples with detailed comments. 

Another point is, I'd like to let users see the options (without duplications) rather than checking API documentation as I am quite sure newbies often misunderstand this. For example, I happen to see newbies setting `inferSchema` to `true` to non-CSV datasources time to time, or setting `com.databricks.spark.csv` instead of `csv`.

**Comment by HyukjinKwon** [comment]
I wasn't even quite sure when I opened the JIRA. That's why I asked it to one of PMCs who might have a better insight. I am okay with going ahead as a small improvement in the docs if any committer likes it (though I don't support) but please leave the JIRA open. I think this PR does not fully solve the issue.

**Comment by felixcheung** [comment]
+1 for more detailed documentation (we should steer away from `inferSchema`)

**Comment by gatorsmile** [comment]
@jomach is a new contributor to Apache Spark. It might be hard for him to address the above comments. Please submit a separate PR for addressing it. Will review it. Thanks!

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

File: docs/sql-programming-guide.md [hunk 14]
```diff
@@ -479,6 +481,26 @@ source type can be converted into other types using this syntax.
...
+<div data-lang="python"  markdown="1">
+{% include_example manual_load_options_csv python/sql/datasource.py %}
+</div>
+
+<div data-lang="r"  markdown="1">
+{% include_example manual_load_options_csv r/RSparkSQLExample.R %}
+
+</div>
+</div>
 ### Run SQL on files directly
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Yup, that's okay. BTW, I initially what I meant in https://github.com/apache/spark/pull/19429#discussion_r143932389 was a newline between `</div>` and `### Run ..` (not `...ample.R %}` and `</div>`. This breaks rendering:




Let's don't forget to fix this up before the release if the followup couldn't be made ahead.

Comment (jomach) [reply docs/sql-programming-guide.md]: @HyukjinKwon  should I add a new line between line 503 and 504 ? 
For example : 
```
{% include_example generic_load_save_functions r/RSparkSQLExample.R %}

</div>
</div>

### Manually Specifying Options
```

Comment (HyukjinKwon) [reply docs/sql-programming-guide.md]: Yup, a newline between 503 and 504.

**Pull request closed without merging**